### DISCUSSION TITLE
Encryption on the server

### DISCUSSION THREAD

**Discussion by BrianHicks** [automerge/automerge#867] [Q&A]
Is there a way to set up automerge so that one peer (a sync server) cannot read data in the document?

I'd like this because I'm writing an app that is single-person but multiple-device. I'd like the server to store data (so it doesn't get erased if indexeddb gets wiped) without being able to read or leak it.

I've searched around a little, but I haven't even found anyone discussing the possibility here.

I'd like to use Automerge for this, but if I need to use a different library I'm also open to recommendations!

**Comment by ept** [comment]
Hi @BrianHicks, this is something we want to support well in the future, but right now it's a bit of a work in progress.

The Automerge sync protocol currently operates on unencrypted data, but you could sync encrypted changes without using the sync protocol, by having the server track which changes it's not yet sent to which client, and vice versa. Then the actual encryption is fairly straightforward, and the bigger issue is key management. The simplest approach to key management would be to have a fixed key per document that you share out-of-band, but that's not ideal for several reasons (to revoke someone's access you'd have to manually rotate the key and distribute the new key to the users; also you don't get authentication of which user made which change).

A more sophisticated solution is @herbcaudill's [local-first-auth](https://github.com/local-first-web/auth). This is probably the direction we want to take end-to-end encryption support in Automerge in the future. But it still needs better integration with automerge-repo and we need better documentation to describe how to work with these libraries.

**Reply by HerbCaudill** [reply]
@BrianHicks @localfirst/auth would give you a mechanism for managing people and their keys - I'd be happy to chat if a walkthrough would be helpful. 

However we don't currently have a way of syncing with a server without giving it visibility into the documents. That's because the easiest way to make a sync server is for it to be "just another peer". @alexjg and I have talked about making a very different sort of sync server, one that stores arbitrary blobs (which could be encrypted operations) along with plaintext dependency information and implements a generic sync protocol that could also work with systems other than Automerge. 

If you'd be interested in helping build something like that, we should talk!

**Reply by BrianHicks** [reply]
Thanks, both! I appreciate the information.

> you could sync encrypted changes without using the sync protocol, by having the server track which changes it's not yet sent to which client, and vice versa.

I had a look at the automerge-repo storage and network adapaters. Which would you recommend using? It seems like some combination of both of those would be necessary to make this work, right? Or would it just be a totally separate thing than automerge-repo at that point?

> However we don't currently have a way of syncing with a server without giving it visibility into the documents. That's because the easiest way to make a sync server is for it to be "just another peer". […] If you'd be interested in helping build something like that, we should talk!

I might be interested! I'd certainly like to *use* something like that. Might as well get the architecture right for the app I'm planning in a way that other folks can use as well. 😁

**Reply by ept** [reply]
Automerge-repo uses the Automerge sync protocol, so I think it won't work with encryption as-is. In principle it would be possible to have a pluggable middleware or something like that in automerge-repo for adding encryption, but I don't think such an API exists yet. So for now I'm afraid you probably have to roll your own. But if you're syncing via a single server, that shouldn't be too hard: after every `Automerge.change`, call `getLastLocalChange()`, which returns a byte array; encrypt it, persist it, and send it to the server (if you're offline right now, enqueue it and send it when you're next online). The server forwards every encrypted change to every other user; when a user receives a change from the server, it decrypts it and calls `applyChange` to apply it to the Automerge doc. That's it, broadly speaking.

**Comment by stephen** [comment]
I haven't tried it, but I happened to stumble on https://github.com/serenity-kit/secsync while researching this topic a while back. Looks like the approach there has some elements of operating on individual updates, but batching them for better perf for some spots.

cc/ @nikgraf perhaps?

**Reply by nikgraf** [reply]
Yes, Secsync is exactly serving this purpose. You can find a working automerge example here https://www.secsync.com/docs/integration-examples/automerge-todos

That said it's not yet ready. There are a few bugs that I want to fix in the coming weeks followed by a Security Scan. If everything goes according to plan we can release a 1.0 in 1-2months 

For more infos: I gave a talk that was quite short notice in London last year. https://www.youtube.com/watch?v=dCGpD8F4isw

**Reply by sachk** [reply]
Looks like secsync would be perfect for me but it hasn't been active for a while. Are there any plans to continue towards a 1.0 at this time?

**Reply by nikgraf** [reply]
@sachk Instead of secsync I would recommend to give KeyHive a try once it's out https://www.inkandswitch.com/keyhive/notebook/

It's a bit early, but here you could even try the code https://github.com/inkandswitch/keyhive
